tmt/steps/prepare/artifact/providers/copr_build.py

```diff
@@ -0,0 +1,216 @@
+"""
+Copr Build Artifact Provider
+"""
+
+import types
+from collections.abc import Sequence
+from functools import cached_property
+from shlex import quote
+from typing import Any, Optional
+from urllib.parse import urljoin
+
+from munch import Munch
+
+import tmt.log
+import tmt.utils
+import tmt.utils.hints
+from tmt.container import container
+from tmt.steps.prepare.artifact import RpmArtifactInfo
+from tmt.steps.prepare.artifact.providers import (
+    ArtifactInfo,
+    ArtifactProvider,
+    ArtifactProviderId,
+    DownloadError,
+    provides_artifact_provider,
+)
+from tmt.steps.provision import Guest
+
+copr: Optional[types.ModuleType] = None
+
+# To silence mypy
+Client: Any
+
+tmt.utils.hints.register_hint(
+    'artifact-provider/copr',
+    """
+The ``copr`` Python package is required by tmt for Copr integration.
+
+To quickly test Copr presence, you can try running:
+
+    python -c 'import copr'
+
+* Users who installed tmt from PyPI should install the ``copr`` package
+  via ``pip install copr``.
+""",
+)
+
+
+def import_copr(logger: tmt.log.Logger) -> None:
+    """Import copr module with error handling."""
+    global copr, Client
+    try:
+        import copr
+        from copr.v3 import Client
+    except ImportError as error:
+        from tmt.utils.hints import print_hints
+
+        print_hints('artifact-provider/copr', logger=logger)
+
+        raise tmt.utils.GeneralError("Could not import copr package.") from error
+
+
+@provides_artifact_provider("copr.build")
+class CoprBuildArtifactProvider(ArtifactProvider[ArtifactInfo]):
+    """
+    Provider for downloading artifacts from Copr builds.
+
+    Identifier format: build-id:chroot-name
+    Example: 1784470:fedora-32-x86_64
+
+    .. note::
+
+        Only RPMs are supported currently.
+
+    .. code-block:: python
+
+        provider = CoprBuildArtifactProvider("copr.build:1784470:fedora-32-x86_64", logger)
+        artifacts = provider.download_artifacts(guest, Path("/tmp"), [])
+    """
+
+    def __init__(self, raw_provider_id: str, logger: tmt.log.Logger):
+        super().__init__(raw_provider_id, logger)
+        self._session = self._initialize_session()
+
+    @cached_property
+    def build_info(self) -> Optional[Munch]:
+        """
+        Fetch and return the build metadata.
+
+        :returns: the build metadata, or ``None`` if not found.
+        """
+        if self.build_id is None:
+            return None
+        build_info = self._call_api("get", self.build_id)
+        assert build_info is None or isinstance(build_info, Munch)
+        return build_info
```

Comment:
Any reason why we are doing `_call_api` for this as well? It would be much easier for the type-checker to use the copr interface directly